### PR TITLE
thread safe set implementation

### PR DESCRIPTION
**Pull request by haridutt12**

### CODE REVIEW

**Comment by haridutt12** [comment]
This is thread safe set implementation along the lines of thread safe Array
fixes #564

**Comment by haridutt12** [comment]
@pitr-ch ping

**Comment by pitr-ch** [comment]
Thank you for your contribution! At first glance it looks good, however I'll have to return to it and give it more thought. Maybe @thedarkone could help. @thedarkone would you have time to help with the review?

File: spec/concurrent/set_spec.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+require 'set' 
+module Concurrent
+  describe Set do 
+    let!(:set) { described_class.new }
+
+    it 'concurrency' do
+      (1..THREADS).map do |i|
+        Thread.new do
+            set << i
```
Comment (thedarkone) [inline spec/concurrent/set_spec.rb]: Change this whole spec to look like `Hash` spec: https://github.com/ruby-concurrency/concurrent-ruby/blob/df482db36caf1b0c1d69a8ff97a2407469e1e315/spec/concurrent/hash_spec.rb#L8 (by adding another inner `1000.times do |j|` loop, just like for `Hash`).

**Comment by haridutt12** [comment]
@thedarkone can u please review the changes.

File: lib/concurrent/set.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    #
+    #   A thread-safe subclass of Set. This version locks against the object
+    #   itself for every method call, ensuring only one thread can be reading
+    #   or writing at a time. This includes iteration methods like `#each`.
+    #
+    #   @note `a += b` is **not** a **thread-safe** operation on
+    #   `Concurrent::Set`. It reads Set `a`, then it creates new `Concurrent::Set`
+    #   which is concatenation of `a` and `b`, then it writes the concatenation to `a`.
+    #   The read and write are independent operations they do not form a single atomic
+    #   operation therefore when two `+=` operations are executed concurrently updates
```
Comment (eregon) [inline lib/concurrent/set.rb]: Unfinished sentence: `updates` ... `may be lost` from https://github.com/ruby-concurrency/concurrent-ruby/blob/df482db36caf1b0c1d69a8ff97a2407469e1e315/lib/concurrent/array.rb#L21

File: spec/concurrent/set_spec.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+module Concurrent
+  describe Set do 
+    let!(:set) { described_class.new }
+
+    it 'concurrency' do
+      (1..THREADS).map do |i|
+        Thread.new do
+           1000.times do |j|
+            set << i 
+            set.empty?
```
Comment (ianks) [inline spec/concurrent/set_spec.rb]: This spec does not make set any expectations. Can you clarify what you are testing and make expectations about the behavior in the spec?

Comment (haridutt12) [reply spec/concurrent/set_spec.rb]: @ianks actually this is my first ever contribution so i dont have much knowledge about minitests so can u plese help me out in this?

**Comment by pitr-ch** [comment]
@thedarkone @ianks @eregon Thanks you for helping with the review!

File: spec/concurrent/set_spec.rb (hunk 3 above)
Comment (pitr-ch) [reply spec/concurrent/set_spec.rb]: Actually the tests follow what we have for Array and Hash. I'll improve all of them at least a bit.

**Comment by pitr-ch** [comment]
I've improved the tests a bit, updated documentation, squashed and rebased. After CI passes it can be merged.

**Pull request merged**